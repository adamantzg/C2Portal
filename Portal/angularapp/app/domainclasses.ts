
export class Product {
    id: number;
    code: string;
    name: string;
    description: string;
    combined_name: string;
}

export class Customer {
    code: string;
    name: string;
    address1: string;
    address2: string;
    address3: string;
    address4: string;
    address5: string;
    address6: string;
    town_city: string;
    county: string;
    analysis_codes1: string;
    currency: string;
    combined_name: string;
}

export class User {
    id: number;
    name: string;
    lastname: string;
    username: string;
    password: string;
    email: string;
    customer_code: string;
    address: string;
    phone: string;
    token: string;
    customer: Customer;
    roles: Role[];
    locked: boolean;
    isTopAdmin: boolean;
    isBranchAdmin: boolean;
    activated: boolean;
    isInternal: boolean;
    lastLogin: Date;
    role: string;
    role_id: number;
    permissions: Permission[];

    constructor() {
        this.customer = new Customer();
    }
}

export class Role {
    id: number;
    name: string;
    permissions: Permission[];
    // strPermissions: string[];

    constructor(id: number, name?: string, permissions?: Permission[]) {
        this.id = id;
        this.name = name;
        this.permissions = permissions;
    }
}


export class Invoice {
    customer: string;
    dated: Date | null;
    kind: string;
    reference: string;
    item_no: string;
    due_date: Date | null;
    amount: number | null;
    open_indicator: string;
    customer_order_num: string;
    selected: boolean;
    postcode: string;
}

export enum RoleId {
    User = 1,
    BranchAdmin = 2,
    Admin = 3
}

export class Order {
    order_no: string; // Sage order number
    customer_order_no: string; // Customer Order number
    date_entered: Date; // Date order received and input to Sage
    invoice_date: Date; // Currently will be to indicate "Delivered" flag. When there is an invoice date then the order is delivered
    date_required: Date;
    address1: string;
    address2: string;
    address3: string;
    address4: string;
    address5: string;
    address6: string;
    town_city: string;
    statusText: string;
    status: number;
    customer?: string;
    country: string;
    audits: OrderAudit[];
    productStockData: StockData[];
    details: OrderDetail[];
    planned: boolean;
}

export class OrderDetail {
    product: string;
    order_qty: number;
    val: number;
    order_line_no: number;
}

export  enum OrderStatus {
    Received = 5,
    InProgress = 7,
    Delivered = 8,
    Cancelled = 9
    // Processed = 'Processed',
    // Confirmed = 'Confirmed',
    // Delivered = 'Delivered',
    // Cancelled = 'Cancelled',
}
export class OrderAudit {
    audit_key: string; // need to join to the Order_no field of opheadm
    character_value: string; // limited to 6 to infer Despatch Note printed
    audit_date: Date; // is the date used for Order Processed date.
}

export class StockData {
    quantity: number;
    product: Product;
    ship_date: Date | null;
    discontinued: boolean;
    madeToOrder: boolean;
    expiresThisYear: boolean;
}

export class Holiday {
    name: string;
    date: Date;
}
export class HolidayExtracted {
    name: string;
    year: number;
    month: number;
    date: number;
    fullDate: Date;
}
export class Permission {
    id: number;
    name: string;
    selected: boolean;

    constructor(id?: number, name?: string, selected?: boolean) {
        this.id = id;
        this.name = name;
        this.selected = selected;
    }
}

export enum PermissionId {
    ViewAccountDetails = 1,
    ViewHolidayAdministration = 2,
    ViewInvoiceHistory = 3,
    ViewOrderHistory = 4,
    ViewStockSearch = 5,
    ViewUserAdministration = 6,
    ViewRoleAdministration = 7
}

export class ClearanceStock {
    product_code: string;
    long_description: string;
    range: string;
    exvat_price: number | null;
    price: number | null;
    freestock: number | null;
    discount: number;
}
