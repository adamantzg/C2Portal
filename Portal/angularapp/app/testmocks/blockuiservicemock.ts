import { EventEmitter } from '@angular/core';

export class BlockUIServiceMock {
    blockTicks = 0;
    unBlockTicks = 0;
    blockUIEvent: any = new EventEmitter<any>();

    startBlock() {
      this.blockTicks++;
    }

    stopBlock() {
      this.unBlockTicks++;
    }
}
