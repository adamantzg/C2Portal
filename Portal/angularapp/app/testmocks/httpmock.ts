import { of } from 'rxjs';

export class HttpMock {
    url: string;
    data: any;
    params: any;
    options: any;
    postResult: any;
    getResult: any;


    post<T>(url, data, params) {
      this.data = data;
      this.url = url;
      this.params = params;
      return of(this.postResult);
    }

    get<T>(url, options) {
        this.url = url;
        this.options = options;
        return of(this.getResult);
    }

  }
