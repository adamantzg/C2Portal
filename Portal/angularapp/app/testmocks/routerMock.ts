import { Component } from '@angular/core';

export class RouterMock {
    navigateUrls: string[];
    params: any;

    navigate(urls: string[], params?: any) {
        this.navigateUrls = urls;
        this.params = params;
    }
}

export class ActivatedRouteMock {

}


export class RouterLinkMock {

}

@Component({
    template: ''
  })
  export class DummyComponent {
  }
