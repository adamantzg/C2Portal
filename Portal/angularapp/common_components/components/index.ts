export * from './grid/grid.component';
export * from './grid/customcolumn/customcolumn.component';
