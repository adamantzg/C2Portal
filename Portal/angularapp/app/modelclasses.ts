import { User, Role, Product, Permission } from './domainclasses';

export class ProductPrices {
    basePrice: number;
    customerPrice: number;
}

export class CustomerTotals  {
    orderTotal: number | null;
    balance: number | null;
    creditLimit: number | null;
    numOfInvoices: number | null;
    numOfOrders: number | null;
}

export class UsersModel {
    users: User[];
    roles: Role[];
}

export class ChangePass {
    code: string;
    username: string;
    password1: string;
    password2: string;
    ruleText: string;
}

export class OrderTotals {
    orderCount: number;
}

export class RolesModel {
    roles: Role[];
    permissions: Permission[];
}


