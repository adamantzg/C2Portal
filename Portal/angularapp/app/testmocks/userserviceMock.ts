import { User } from '../domainclasses';

export class UserServiceMock {
    CurrentUser: User;
}
