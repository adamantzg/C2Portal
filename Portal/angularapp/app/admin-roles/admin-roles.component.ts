import { Component, OnInit } from '@angular/core';
import { UserService } from '../services/user.service';
import { Role, Permission, RoleId } from '../domainclasses';
import { GridDefinition, GridColumn, GridButton, GridColumnType, GridEditorType,
    ColumnDataType, GridEditMode, GridButtonEventData, MultipleSelectionData } from '../../common_components/components';
import { RolesModel } from '../modelclasses';
import { MessageboxService } from '../common/messagebox/messagebox.service';
import { MessageBoxType, MessageBoxCommand, MessageBoxCommandValue } from '../common/ModalDialog';
import { CommonService } from '../../c2-flex/services/common.service';

@Component({
    selector: 'admin-roles',
    templateUrl: './admin-roles.component.html',
    styleUrls: ['./admin-roles.component.css']
  })
export class AdminRolesComponent implements OnInit {
    constructor(private userService: UserService, private messageBoxService: MessageboxService,
    private commonService: CommonService) {
        this.gridDefinition.columnButtonVisibilityCallback = this.checkColumnButtonVisibility.bind(this);
    }
    model: RolesModel = new RolesModel();
    errorMessage = '';
    successMessage = '';
    editModes = Object.assign({}, GridEditMode);
    editMode = GridEditMode.NoEdit;
    roleCopy: Role;
    editedRoleId = -1;

    gridDefinition = new GridDefinition([
        new GridColumn('Name', 'name', GridColumnType.Label, 'name', null, null, null, null, true),
        new GridColumn('Permissions', 'permissions', GridColumnType.Label, 'permissions', null, null, null, null, true, null,
        null, null, true, GridEditorType.MultipleChoiceCheckboxes, null, null, true, null,
        new MultipleSelectionData(true, null, 'id', 'name', 'selected', 'Choose permission'),
         this.permissionsString ),
        new GridColumn('buttons', null, GridColumnType.ButtonGroup, 'buttons', null, null, null, null,
        false, null, null, null, false,
          null,
          [
            new GridButton('Update', 'update', 'fa fa-check', 'btn-xs', true),
            new GridButton('Cancel', 'cancel', 'fa-times-circle-o', 'btn-xs'),
            new GridButton('Edit', 'edit', 'fa-pencil-square-o', 'btn-xs'),
            new GridButton('Delete', 'delete', 'fa-remove', 'btn-danger btn-xs')
          ])
    ]);

    ngOnInit() {
        this.userService.getRolesModel().subscribe((m: RolesModel) => {
            this.model = m;
            // this.model.roles.forEach(r => r.strPermissions = r.permissions.map(p => p.name));
        });
    }

    addNew() {
        this.roleCopy = JSON.parse(JSON.stringify(new Role(0)));
        this.roleCopy.permissions = JSON.parse(JSON.stringify(this.model.permissions));
        this.editMode = GridEditMode.AddNew;
        this.editedRoleId = 0;
    }

    gridButtonClicked(event: GridButtonEventData) {
        this.clearMessages();
        if (event.name === 'delete') {
          this.onDeleteRow(event.row);
        } else if (event.name === 'edit') {
          this.onEditRow(event.row);
          this.editMode = GridEditMode.Edit;
        } else if (event.name === 'cancel') {
          this.editMode = GridEditMode.NoEdit;
        } else if (event.name === 'update') {
          this.onUpdateRow();
        }
    }

    permissionsString(data: Role) {
        if (data.permissions != null) {
            return data.permissions.sort( (x, y) => x.id - y.id).map(p => p.name).join(', ');
        }
        return '';
    }

    clearMessages() {
        this.errorMessage = '';
        this.successMessage = '';
    }

    checkColumnButtonVisibility(r: Role, column: string, button: string) {

        if (button === 'edit' || button === 'delete') {
          return (r.id !== RoleId.Admin) &&  (this.editMode === GridEditMode.NoEdit || r.id !== this.roleCopy.id);
        } else if (button === 'update' || button === 'cancel') {
          return this.editMode !== GridEditMode.NoEdit && r.id === this.roleCopy.id;
        }
        return true;
    }

    onEditRow(role: Role) {
        this.roleCopy = JSON.parse(JSON.stringify(role));
        const selectedPermissions = {};
        this.roleCopy.permissions.forEach(p => selectedPermissions[p.id] = true);
        this.roleCopy.permissions = JSON.parse(JSON.stringify(this.model.permissions));
        this.roleCopy.permissions.forEach(p => p.selected = selectedPermissions[p.id] != null );
        this.editedRoleId = role.id;
      }

      onCancelRow() {
        this.editedRoleId = -1;
      }
      onUpdateRow() {
        /** send user to services */

        const role = JSON.parse(JSON.stringify(this.roleCopy));
        const isNew = role.id === 0;
        this.errorMessage = '';
        role.permissions = role.permissions.filter(p => p.selected);

        // tslint:disable-next-line:triple-equals

        this.userService.createUpdateRole(role).subscribe( (r: Role) => {
          this.errorMessage = '';

          if (isNew) {
            this.model.roles.push(r);
          } else {
            const existingRole = this.model.roles.find(x => x.id === role.id);
            if (existingRole != null) {
              Object.assign(existingRole, r);
            }

          }
          this.editMode = GridEditMode.NoEdit;
            },
            err => this.errorMessage = this.commonService.getError(err)
        );

      }
      onDeleteRow(role: Role) {
        this.messageBoxService.openDialog('Delete role?', MessageBoxType.Yesno).subscribe((m: MessageBoxCommand) => {
          if (m.value === MessageBoxCommandValue.Ok) {
            this.errorMessage = '';
            this.userService.deleteRole(role.id).subscribe( data => {
              const index = this.model.roles.findIndex(r => r.id === role.id);
              if (index >= 0) {
                this.model.roles.splice(index, 1);
              }
            },
            err => this.errorMessage = this.commonService.getError(err));
          }
        });

      }
}


