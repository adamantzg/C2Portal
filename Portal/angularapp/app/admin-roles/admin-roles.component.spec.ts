import { UserService } from '../services/user.service';
import { HttpClient } from '@angular/common/http';
import { HttpMock } from '../testmocks/httpmock';
import { CommonAppModule } from 'common_components/common.module';
import { BlockUIService } from 'common_components/services/block-ui.service';
import { BlockUIServiceMock } from '../testmocks/blockuiservicemock';
import { ComponentFixture, TestBed } from '@angular/core/testing';
import { AdminRolesComponent } from './admin-roles.component';
import { User, Role, Permission, PermissionId, RoleId } from '../domainclasses';
import { RolesModel } from '../modelclasses';
import { of } from 'rxjs';
import { BlockUiComponent } from '../blockui/block-ui.component';
import { RouterTestingModule } from '@angular/router/testing';
import { HttpService } from '../services/http.service';
import { MessageboxService } from '../common/messagebox/messagebox.service';
import { CommonService } from '../../c2-flex/services/common.service';
import { BsModalService, ModalModule } from 'ngx-bootstrap';
import { MessageBoxCommand, MessageBoxCommandValue } from '../common/ModalDialog';
import { GridButtonEventData, GridEditMode } from 'common_components/components';



describe('admin-roles.component', () => {

    let fixture: ComponentFixture<AdminRolesComponent>;
    let component: AdminRolesComponent;
    let userService: UserService;
    let model: RolesModel;

    setupTestBed({
        declarations: [AdminRolesComponent, BlockUiComponent],
        imports: [CommonAppModule, RouterTestingModule.withRoutes([{path: 'user', component: AdminRolesComponent}]), ModalModule.forRoot()],
        providers: [UserService, HttpService, {provide: HttpClient, useClass: HttpMock},
            {provide: BlockUIService, useClass: BlockUIServiceMock}, MessageboxService, CommonService, BsModalService]
    });

    beforeEach(() => {
        fixture = TestBed.createComponent(AdminRolesComponent);
        component = fixture.componentInstance;

        userService = TestBed.get(UserService);
        const u = new User();
        u.id = 1;
        u.isTopAdmin = true;
        model = new RolesModel();

        jest.spyOn(userService, 'CurrentUser', 'get').mockImplementation(() => u);
        jest.spyOn(userService, 'getRolesModel').mockImplementation(() => of(model));
    });

    test('should create', () => {
        expect(component).toBeTruthy();
        expect(fixture).toMatchSnapshot();
    });

    test('checkColumnButtonVisibility', () => {

        const r = new Role(3, 'admin');
        let result = component.checkColumnButtonVisibility(r, '', 'edit');
        expect(result).toBeFalsy();

        r.id = 4;
        result = component.checkColumnButtonVisibility(r, '', 'edit');
        expect(result).toBeTruthy();
    });

    test('onEditRow', () => {

        model.permissions = [new Permission(PermissionId.ViewAccountDetails), new Permission(PermissionId.ViewInvoiceHistory)] ;
        component.model = model;
        const r = new Role(4);

        r.permissions = [new Permission(PermissionId.ViewAccountDetails)];
        component.onEditRow(r);
        expect(component.roleCopy).toBeTruthy();
        expect(component.roleCopy.id).toBe(r.id);

        expect(component.roleCopy.permissions.length).toBe(2);
        expect(component.roleCopy.permissions.filter(p => p.selected).length).toBe(1);
    });

    test('onUpdateRow', () => {
        model.roles = [new Role(RoleId.User), new Role(RoleId.BranchAdmin)] ;
        component.model = model;
        const r = new Role(0);
        r.permissions = [new Permission(PermissionId.ViewAccountDetails), new Permission(PermissionId.ViewInvoiceHistory)];
        r.permissions[0].selected = true;
        component.roleCopy = r;
        const returnedRole = new Role(4);

        const fn = jest.spyOn(userService, 'createUpdateRole').mockImplementation(() => of(returnedRole));

        component.onUpdateRow();
        expect(fn.mock.calls[0][0].permissions.length).toBe(1);
        expect(model.roles.length).toBe(3);

        // Check existing
        r.id = RoleId.User;
        returnedRole.name = 'changed';
        component.onUpdateRow();
        expect(model.roles[0].name).toBe(returnedRole.name);

    });

    test('onDeleteRow', () => {
        component.model.roles = [new Role(RoleId.User), new Role(RoleId.BranchAdmin)];
        const msgBoxService = TestBed.get(MessageboxService);
        const cmd = new MessageBoxCommand(MessageBoxCommandValue.Ok);
        const fn1 = jest.spyOn(msgBoxService, 'openDialog').mockImplementation(() => of(cmd));
        const fn2 = jest.spyOn(userService, 'deleteRole').mockImplementation(() => of(null));
        component.onDeleteRow(new Role(RoleId.User));
        expect(fn1.mock.calls[0][0]).toBe('Delete role?');
        expect(fn2.mock.calls[0][0]).toBe(RoleId.User);
        expect(component.model.roles.length).toBe(1);
    });

    test('permissionsString', () => {
        const data = new Role(4);
        data.permissions = [new Permission(PermissionId.ViewInvoiceHistory, 'view invoice'),
        new Permission(PermissionId.ViewAccountDetails, 'view account')];
        const arrResult = component.permissionsString(data).split(', ');
        expect(arrResult.length).toBe(2);
        expect(arrResult[0]).toBe('view account');
    });

    test('gridButtonClicked', () => {
        const data = new GridButtonEventData('c', {id: 1});
        const fn1 = jest.spyOn(component, 'onDeleteRow').mockImplementation(() => {});
        let fn2 = jest.spyOn(component, 'onEditRow').mockImplementation(() => {});
        data.name = 'delete';
        component.gridButtonClicked(data);
        expect(fn1.mock.calls.length).toBe(1);
        expect(fn1.mock.calls[0][0].id).toBe(data.row.id);
        data.name = 'edit';
        component.gridButtonClicked(data);
        expect(fn2.mock.calls.length).toBe(1);
        expect(fn2.mock.calls[0][0].id).toBe(data.row.id);
        expect(component.editMode).toBe(GridEditMode.Edit);
        data.name = 'cancel';
        component.gridButtonClicked(data);
        expect(component.editMode).toBe(GridEditMode.NoEdit);
        data.name = 'update';
        fn2 = jest.spyOn(component, 'onUpdateRow').mockImplementation(() => {});
        component.gridButtonClicked(data);
        expect(fn2.mock.calls.length).toBe(1);

    });
});
