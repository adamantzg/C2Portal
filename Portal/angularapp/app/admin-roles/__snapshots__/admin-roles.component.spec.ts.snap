// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin-roles.component should create 1`] = `
<admin-roles
  commonService={[Function CommonService]}
  editMode="0"
  editModes={[Function Object]}
  editedRoleId={[Function Number]}
  errorMessage=""
  gridDefinition={[Function GridDefinition]}
  messageBoxService={[Function MessageboxService]}
  model={[Function RolesModel]}
  successMessage=""
  userService={[Function UserService]}
>
  <div
    class="container-fluid"
    style="height: 100%"
  >
    
    
    <div
      class="row admin-row"
    >
      <div
        class="col-sm-12"
      >
        <br />
        <button
          class="btn btn-default"
        >
          <i
            aria-hidden="true"
            class="fa fa-plus-square-o"
          />
           Add new 
        </button>
      </div>
    </div>
    <div
      class="col-sm-12"
    />
    <div
      class="row"
      style="height: 100%"
    >
      <div
        class="col-sm-12"
        style="height: 100%"
      >
        <app-grid>
          <form
            novalidate=""
          >
            <div>
              <table
                style="margin-bottom: 0px"
              >
                <thead>
                  <tr>
                    
                  </tr>
                </thead>
                
              </table>
              
              
              
            </div>
          </form>
        </app-grid>
      </div>
    </div>
  </div><app-block-ui>
    
  </app-block-ui>
</admin-roles>
`;
